src/phoenix/server/api/mutations/dataset_split_mutations.py

```diff
@@ -0,0 +1,297 @@
+            existing_dataset_example_split_keys = await session.execute(
+                select(
+                    models.DatasetSplitDatasetExample.dataset_split_id,
+                    models.DatasetSplitDatasetExample.dataset_example_id,
+                ).where(
+                    models.DatasetSplitDatasetExample.dataset_split_id.in_(dataset_split_ids)
+                    & models.DatasetSplitDatasetExample.dataset_example_id.in_(example_ids)
+                )
+            )
+            unique_dataset_example_split_keys = set(existing_dataset_example_split_keys.all())
+
+            # Compute all desired pairs and insert only missing
+            values = []
+            for dataset_split_id in dataset_split_ids:
+                for example_id in example_ids:
+                    # if the keys already exists, skip
+                    if (dataset_split_id, example_id) in unique_dataset_example_split_keys:
+                        continue
+                    dataset_split_id_key = models.DatasetSplitDatasetExample.dataset_split_id.key
+                    values.append(
+                        {
+                            dataset_split_id_key: dataset_split_id,
+                            models.DatasetSplitDatasetExample.dataset_example_id.key: example_id,
+                        }
+                    )
+
+            if values:
+                try:
+                    await session.execute(insert(models.DatasetSplitDatasetExample), values)
+                    await session.flush()
+                except (PostgreSQLIntegrityError, SQLiteIntegrityError) as e:
+                    raise Conflict("Failed to add examples to dataset splits.") from e
+
+        return AddDatasetExamplesToDatasetSplitsMutationPayload(
+            query=Query(),
+        )
+
+    @strawberry.mutation(permission_classes=[IsNotReadOnly])  # type: ignore
+    async def remove_dataset_examples_from_dataset_splits(
+        self, info: Info[Context, None], input: RemoveDatasetExamplesFromDatasetSplitsInput
+    ) -> RemoveDatasetExamplesFromDatasetSplitsMutationPayload:
+        async with info.context.db() as session:
+            dataset_split_ids = [
+                from_global_id_with_expected_type(dataset_split_id, DatasetSplit.__name__)
+                for dataset_split_id in input.dataset_split_ids
+            ]
+            example_ids = [
+                from_global_id_with_expected_type(example_id, models.DatasetExample.__name__)
+                for example_id in input.example_ids
+            ]
+            if not example_ids:
+                raise Conflict("No examples provided.")
+
+            stmt = delete(models.DatasetSplitDatasetExample).where(
+                models.DatasetSplitDatasetExample.dataset_split_id.in_(dataset_split_ids)
+                & models.DatasetSplitDatasetExample.dataset_example_id.in_(example_ids)
+            )
+
+            await session.execute(stmt)
```

Comment:
fwiw. only open the transaction when everything is ready
```suggestion
    async with info.context.db() as session
            await session.execute(stmt)
```